In [33]:
import numpy as np
import matplotlib.pyplot as plt

# PART A

In [34]:
np.random.seed(0)
x = np.linspace(-5, 5, 50)
y = 2 * x**2 + 3 * x + np.random.randn(50) * 10

In [35]:
def mse(y_true, y_pred):
    return np.mean((y_true - y_pred)**2)

def r_squared(y_true, y_pred):
    ss_total = np.sum((y_true - np.mean(y_true))**2)
    ss_res = np.sum((y_true - y_pred)**2)
    return 1 - (ss_res/ss_total)

In [36]:
A = np.vstack([x, np.ones(len(x))]).T
m,c = np.linalg.lstsq(A,y,rcond=None)[0]
y_ls = m*x + c
mse_ls = mse(y,y_ls)
r2_ls = r_squared(y,y_ls)
print("MSE:", mse_ls)
print("R^2:", r2_ls)

MSE: 372.0051625579934
R^2: 0.05555558712253694


In [37]:
degree = 3

coeffs = np.polyfit(x,y,degree)
poly = np.poly1d(coeffs)

y_poly = poly(x)

mse_poly = mse(y,y_poly)
r2_poly = r_squared(y,y_poly)

print("MSE:", mse_poly)
print("R^2:", r2_poly)

MSE: 107.11173673839265
R^2: 0.7280653832313249


## PART A DISCUSSION

The polynomial fit is definitely the better fit due to the higher r-squared value as well as the much lower MSE value. The 372 MSE and .05 R^2 for the lstsqr shows that it is clearly not a good fit for the data points. On the other hand, the poly has a R^2 value of .728 which is in the perfect range to depict a good fit. 

# PART B

In [44]:
for deg in range(1,37):

    coeffs = np.polyfit(x,y,deg)
    poly = np.poly1d(coeffs)

    y_pred = poly(x)

    m = mse(y,y_pred)
    r2 = r_squared(y,y_pred)

    print(f"Degree {deg}: MSE={m:.3f}, R^2={r2:.3f}")

Degree 1: MSE=372.005, R^2=0.056
Degree 2: MSE=108.886, R^2=0.724
Degree 3: MSE=107.112, R^2=0.728
Degree 4: MSE=106.646, R^2=0.729
Degree 5: MSE=106.561, R^2=0.729
Degree 6: MSE=106.508, R^2=0.730
Degree 7: MSE=105.983, R^2=0.731
Degree 8: MSE=105.680, R^2=0.732
Degree 9: MSE=103.545, R^2=0.737
Degree 10: MSE=103.472, R^2=0.737
Degree 11: MSE=102.277, R^2=0.740
Degree 12: MSE=90.960, R^2=0.769
Degree 13: MSE=90.893, R^2=0.769
Degree 14: MSE=90.674, R^2=0.770
Degree 15: MSE=86.170, R^2=0.781
Degree 16: MSE=86.067, R^2=0.781
Degree 17: MSE=84.824, R^2=0.785
Degree 18: MSE=78.567, R^2=0.801
Degree 19: MSE=60.532, R^2=0.846
Degree 20: MSE=60.297, R^2=0.847
Degree 21: MSE=59.052, R^2=0.850
Degree 22: MSE=58.669, R^2=0.851
Degree 23: MSE=49.883, R^2=0.873
Degree 24: MSE=49.443, R^2=0.874
Degree 25: MSE=44.908, R^2=0.886
Degree 26: MSE=44.229, R^2=0.888
Degree 27: MSE=44.125, R^2=0.888
Degree 28: MSE=42.752, R^2=0.891
Degree 29: MSE=42.566, R^2=0.892
Degree 30: MSE=35.209, R^2=0.911
Degree 3

/opt/conda/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3326: RankWarning: Polyfit may be poorly conditioned
  exec(code_obj, self.user_global_ns, self.user_ns)


Here we can see that the polynomial begins to blatantly overfit at around degree 36. It is most likely earlier than this, however it is at this point that we begin to recieve errors. On the other side, degree 1 seems to be the most liklely candidate for the underfit. It is at this point that the R^2 value drops immensely and the MSE shoots up.

# PART C

In [46]:
degree = 3

V = np.vander(x, degree+1)

coeffs_v = np.linalg.lstsq(V,y,rcond=None)[0]

y_vander = V @ coeffs_v

mse_v = mse(y,y_vander)
r2_v = r_squared(y,y_vander)

print("\nVandermonde Polynomial Fit")
print("MSE:", mse_v)
print("R^2:", r2_v)


Vandermonde Polynomial Fit
MSE: 107.11173673839266
R^2: 0.7280653832313249
